Adopt more smart pointers in NetworkStorageManager, SQLiteStorageArea

jelee53 opened the pull request:
#### 997ad3ca4c323f74c8d3575eacce8b7088575885
<pre>
Adopt more smart pointers in NetworkStorageManager, SQLiteStorageArea
<a href="https://bugs.webkit.org/show_bug.cgi?id=295612">https://bugs.webkit.org/show_bug.cgi?id=295612</a>
<a href="https://rdar.apple.com/155384272">rdar://155384272</a>

Reviewed by Per Arne Vollan and Ryosuke Niwa.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.</a>

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::estimate):
(WebKit::NetworkStorageManager::fetchSessionStorageForWebPage):
(WebKit::NetworkStorageManager::setOriginQuotaRatioEnabledForTesting):
(WebKit::NetworkStorageManager::cancelConnectToStorageArea):
(WebKit::NetworkStorageManager::unlockCacheStorage):
(WebKit::NetworkStorageManager::cacheStorageClearMemoryRepresentation):
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
(WebKit::SQLiteStorageArea::isEmpty):
(WebKit::SQLiteStorageArea::createTableIfNecessary):
(WebKit::SQLiteStorageArea::prepareDatabase):
(WebKit::SQLiteStorageArea::startTransactionIfNecessary):
(WebKit::SQLiteStorageArea::cachedStatement):
(WebKit::SQLiteStorageArea::checkedDatabase const):
(WebKit::SQLiteStorageArea::getItemFromDatabase):
(WebKit::SQLiteStorageArea::allItems):
(WebKit::SQLiteStorageArea::setItem):
(WebKit::SQLiteStorageArea::removeItem):
(WebKit::SQLiteStorageArea::clear):
(WebKit::SQLiteStorageArea::handleLowMemoryWarning):
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.h:
* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:

Canonical link: <a href="https://commits.webkit.org/297257@main">https://commits.webkit.org/297257@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d06bcf0e24c4533a8e5678a338f99a3b55329c77

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111071 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30737 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21168 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117102 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61338 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31418 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39319 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84440 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114018 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25101 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99999 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64886 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24444 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60923 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94480 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18210 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119963 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38120 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28328 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93378 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38496 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96277 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93202 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38284 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16025 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34101 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17913 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38009 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43485 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37673 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41007 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39376 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jelee53 commented on Source/WebKit/NetworkProcess/storage/SessionStorageManager.cpp:
```diff
@@ -73,13 +73,13 @@ void SessionStorageManager::removeNamespace(StorageNamespaceIdentifier namespace
 {
     auto identifier = storageArea->identifier();
-    m_registry->registerStorageArea(identifier, storageArea);
+    CheckedRef { m_registry }->registerStorageArea(identifier, storageArea);
```
TODO: Remove this file because Anne's PR will fix it for being `const`. This is the incorrect approach to fix it.

annevk commented on Source/WebCore/platform/sql/SQLiteStatementAutoResetScope.h:
```diff
@@ -46,6 +46,7 @@ class SQLiteStatementAutoResetScope {
+    WEBCORE_EXPORT CheckedPtr<SQLiteStatement> checkedGet();
 
 private:
     CheckedPtr<SQLiteStatement> m_statement;
```
It seems you could make this const instead?

jelee53 replied:
Hi Anne - I tried this and it doesn't work :( 

```
In file included from /Users/jessicalee/Documents/webkit-col/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource391.cpp:7:
/Users/jessicalee/Documents/webkit-col/OpenSource/Source/WebCore/platform/sql/ :43:63: error: defaulting this move assignment operator would delete it after its first declaration
   43 | SQLiteStatementAutoResetScope& SQLiteStatementAutoResetScope::operator=(SQLiteStatementAutoResetScope&&) = default;
In file included from /Users/jessicalee/Documents/webkit-col/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource391.cpp:7:
In file included from /Users/jessicalee/Documents/webkit-col/OpenSource/Source/WebCore/platform/sql/SQLiteStatementAutoResetScope.cpp:27:
/Users/jessicalee/Documents/webkit-col/OpenSource/Source/WebCore/platform/sql/SQLiteStatementAutoResetScope.h:51:39: note: move assignment operator of 'SQLiteStatementAutoResetScope' is implicitly deleted because field 'm_statement' has no move assignment operator
   51 |     const CheckedPtr<SQLiteStatement> m_statement;
1 error generated.
```

Basically, what it's saying is that SQLiteStatementAutoResetScope overrides the move assignment. 
```
SQLiteStatementAutoResetScope& SQLiteStatementAutoResetScope::operator=(SQLiteStatementAutoResetScope&&) = default;
```

If it's const, you can't do a move-assignment after m_statement is created. 

So if we want to do this, 
```
SQLiteStatementAutoResetScope foo = <..instantiated.. >; 
SQLiteStatementAutoResetScope bar =< .. instantiated .. >; 
foo = bar; 
```

making the m_statement member `const` won't allow us to do that. Because the member is `const` we can't do something like 
```
this->m_statement = WTFMOVE(bar.m_statement); // part of the SQLiteStatementAutoResetScope move-assignment override
```

So we need to leave it non-const.

rniwa commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -119,7 +119,7 @@ bool SQLiteStorageArea::isEmpty()
         return true;
 
-    auto statement = cachedStatement(StatementType::CountItems);
+    CheckedPtr statement = cachedStatement(StatementType::CountItems).checkedGet();
```
You use regular `get()` when you're specifying the type on the LHS like this.

rniwa commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -266,7 +274,7 @@ Expected<String, StorageError> SQLiteStorageArea::getItemFromDatabase(const Stri
     // Ensure that statement goes out of scope before handleDatabaseErrorIfNeeded(). Otherwise, we may get a CheckedPtr verification failure.
     {
-        auto statement = cachedStatement(StatementType::GetItem);
+        CheckedPtr statement = cachedStatement(StatementType::GetItem).checkedGet();
```
Ditto about using `get()` instead.

rniwa commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -327,7 +335,7 @@ HashMap<String, String> SQLiteStorageArea::allItems()
     // Ensure that statement goes out of scope before handleDatabaseErrorIfNeeded(). Otherwise, we may get a CheckedPtr verification failure.
     {
-        auto statement = cachedStatement(StatementType::GetAllItems);
+        CheckedPtr statement = cachedStatement(StatementType::GetAllItems).checkedGet();
```
Ditto about using `get()` instead.

rniwa commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -379,7 +387,7 @@ Expected<void, StorageError> SQLiteStorageArea::setItem(std::optional<IPC::Conne
     // Ensure that statement goes out of scope before handleDatabaseErrorIfNeeded(). Otherwise, we may get a CheckedPtr verification failure.
     {
-        auto statement = cachedStatement(StatementType::SetItem);
+        CheckedPtr statement = cachedStatement(StatementType::SetItem).checkedGet();
```
Ditto about using `get()` instead.

rniwa commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -421,7 +429,7 @@ Expected<void, StorageError> SQLiteStorageArea::removeItem(IPC::Connection::Uniq
     // Ensure that statement goes out of scope before handleDatabaseErrorIfNeeded(). Otherwise, we may get a CheckedPtr verification failure.
     {
-        auto statement = cachedStatement(StatementType::DeleteItem);
+        CheckedPtr statement = cachedStatement(StatementType::DeleteItem).checkedGet();
```
Ditto about using `get()` instead.

rniwa commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -466,7 +474,7 @@ Expected<void, StorageError> SQLiteStorageArea::clear(IPC::Connection::UniqueID
     // Ensure that statement goes out of scope before handleDatabaseErrorIfNeeded(). Otherwise, we may get a CheckedPtr verification failure.
     {
-        auto statement = cachedStatement(StatementType::DeleteAllItems);
+        CheckedPtr statement = cachedStatement(StatementType::DeleteAllItems).checkedGet();
```
Ditto about using `get()` instead.

cdumez commented on Source/WebCore/platform/sql/SQLiteStatementAutoResetScope.cpp:
```diff
@@ -39,6 +39,11 @@ SQLiteStatementAutoResetScope::SQLiteStatementAutoResetScope(SQLiteStatement *st
 {
 }
 
+CheckedPtr<SQLiteStatement> SQLiteStatementAutoResetScope::checkedGet()
```
We do not need this function.

jelee53 commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1616,17 +1616,19 @@ void NetworkStorageManager::cancelConnectToStorageArea(IPC::Connection& connecti
         return;
 
     auto connectionIdentifier = connection.uniqueID();
+    auto originStorageManager = CheckedRef { *(iterator->value) };
```
I'll change this to `CheckedRef originStorageManager =  *(iterator->value)` too. My bad.

pvollan commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -165,7 +166,8 @@ bool SQLiteStorageArea::createTableIfNecessary()
 {
-    if (m_database && m_database->isOpen())
+    CheckedPtr maybeOpenDatabase = m_database.get();
+    if (maybeOpenDatabase && maybeOpenDatabase->isOpen())
```
Nit: perhaps we can just use the name `database`?

pvollan reviewed (approved):
R=me.

jelee53 committed:
Adopt more smart pointers in NetworkStorageManager, SQLiteStorageArea
https://bugs.webkit.org/show_bug.cgi?id=295612
rdar://155384272

Reviewed by Per Arne Vollan and Ryosuke Niwa.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::estimate):
(WebKit::NetworkStorageManager::fetchSessionStorageForWebPage):
(WebKit::NetworkStorageManager::setOriginQuotaRatioEnabledForTesting):
(WebKit::NetworkStorageManager::cancelConnectToStorageArea):
(WebKit::NetworkStorageManager::unlockCacheStorage):
(WebKit::NetworkStorageManager::cacheStorageClearMemoryRepresentation):
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
(WebKit::SQLiteStorageArea::isEmpty):
(WebKit::SQLiteStorageArea::createTableIfNecessary):
(WebKit::SQLiteStorageArea::prepareDatabase):
(WebKit::SQLiteStorageArea::startTransactionIfNecessary):
(WebKit::SQLiteStorageArea::cachedStatement):
(WebKit::SQLiteStorageArea::checkedDatabase const):
(WebKit::SQLiteStorageArea::getItemFromDatabase):
(WebKit::SQLiteStorageArea::allItems):
(WebKit::SQLiteStorageArea::setItem):
(WebKit::SQLiteStorageArea::removeItem):
(WebKit::SQLiteStorageArea::clear):
(WebKit::SQLiteStorageArea::handleLowMemoryWarning):
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.h:
* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:

Canonical link: https://commits.webkit.org/297257@main
